Add INCIDENT_RESTART event

jladd-geant opened the pull request:
## Scope and purpose

Fixes #1502. <!-- Reference: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->

Depends on #1506.

Adds a new INCIDENT_RESTART event to allow source systems to reopen incidents.
Described further in linked issue.

<!-- remove things that do not apply -->
### This pull request
* changes the database


## Contributor Checklist

Every pull request should have this checklist filled out, no matter how small it is.
More information about contributing to Argus can be found in the
[Development docs](https://argus-server.readthedocs.io/en/latest/development.html).

<!-- Add an "X" inside the brackets to confirm -->
<!-- Remove checks that do not apply -->
<!-- Of the checks that do apply: If not checking one or more of the boxes, please explain why below each. -->

* [x] Added a changelog fragment for [towncrier](https://argus-server.readthedocs.io/en/latest/development/howtos/changelog-entry.html)
* [x] Added/amended tests for new/changed code
<del>Doesn't break any tests, doesn't introduce any untested lines of code, and is implemented in the same way as the existing REOPEN event. So adding/changing tests didn't seem necessary, happy to do otherwise if needed.</del>

* [x] Added/changed documentation, including updates to the [user manual](https://argus-server.readthedocs.io/en/latest/user-manual.html) if feature flow or UI is considerably changed
* [x] Linted/formatted the code with ruff and djLint, easiest by using [pre-commit](https://github.com/Uninett/Argus?tab=readme-ov-file#code-style)
* [x] The first line of the commit message continues the sentence "If applied, this commit will ...", starts with a capital letter, does not end with punctuation and is 50 characters or less long. See our [how-to](https://argus-server.readthedocs.io/en/latest/development/howtos/commit-messages.html)
* [ ] If this results in changes to the database model: <del>Updated the [ER diagram](https://argus-server.readthedocs.io/en/latest/development/howtos/regenerate-the-ER-diagram.html)</del>
There are no changes to the ER diagram as a result of adding a new event type

<!-- Make this a draft PR if the content is subject to change, cannot be merged or if it is for initial feedback -->

jladd-geant commented:
Only thing im unsure of is the migration generated seems to be duplicating changes? It seems to be converting the id of some fields to BigAutoField despite that already being in the previous squashed migration? I cant imagine it would cause any issues but not really sure whats going on with that.
Let me know if I should make any changes.

hmpf commented:
Agree about the migrations, the change to BigAutoField should already have happened . On Tuesday I hope to be able to check this properly.

Do `python manage.py dbshell` against prod and dev database. In dbshell run `select * from django_migrations where app = 'argus_incident';`.

You should have this line (first column may be different, last column deffo different)

```
 89 | argus_incident | 0009_use_bigautofield_on_quickly_growing_tables | 2025-05-16 08:23:36.499823+02
```

If not I'll see if I can do some experiments (with your help) next week, we might need a release that canonicalizes the squashed migrations (=deletes the old ones, needs manual rewrite of the new migrations).

Also do `\d argus_incident_incident` and check that the `Type` of the `id` field is already `bigint`:

```
       Column       |           Type           | Collation | Nullable |                       Default                       
--------------------+--------------------------+-----------+----------+-----------------------------------------------------
 id                 | bigint                   |           | not null | nextval('argus_incident_incident_id_seq'::regclass)
```

hmpf commented on src/argus/incident/views.py:
```diff
@@ -489,7 +489,7 @@ def update_incident(self, validated_data: dict, incident: Incident):
             incident.end_time = timestamp
             incident.save()
-        elif event_type == Event.Type.REOPEN:
+        elif event_type in {Event.Type.REOPEN, Event.Type.INCIDENT_RESTART} :
```
For a later branch, I might already have an outdated one, we should make some constants for the sets used here, makes for better inline docs.

johannaengland replied:
And in line 467 `if event_type in {Event.Type.INCIDENT_START, Event.Type.INCIDENT_END}:` you need to remove INCIDENT_END from the check, otherwise you will be able to restart an event, but not end it again through the source system

jladd-geant replied:
Fixed in 2fc6736e5ce769726a9c829c0ac2b09e14c5d256

johannaengland commented on docs/reference/api/v2.rst:
```diff
@@ -393,6 +393,12 @@ Incident endpoints
         the incident if it has been closed (either manually or by a
         source system).
 
+      -  ``RES`` - Incident restart
```
In the lines 382/383 it says `An incident cannot have more than one event of this type.` this needs to be removed as well.

johannaengland reviewed (changes requested):
I agree both with @hmpf  and @elfjes, looking into the migration problem is important and I would like to see more tests. 

And I have a few additional comments, because currently a source system can restart an incident, but not end it again

jladd-geant commented:
> Agree about the migrations, the change to BigAutoField should already have happened . On Tuesday I hope to be able to check this properly.
> 
> Do `python manage.py dbshell` against prod and dev database. In dbshell run `select * from django_migrations where app = 'argus_incident';`.
> 
> You should have this line (first column may be different, last column deffo different)
> 
> ```
>  89 | argus_incident | 0009_use_bigautofield_on_quickly_growing_tables | 2025-05-16 08:23:36.499823+02
> ```
> 
> If not in prod db I'll see if I can do some experiments (with your help) next week, we might need a release that canonicalizes the squashed migrations (=deletes the old ones, needs manual rewrite of the new migrations). If not in dev: nuke the dev database and migrate to version 1.37.0 and check again before you migrate to 2.0.0.
> 
> Also do `\d argus_incident_incident` and check that the `Type` of the `id` field is already `bigint`:
> 
> ```
>        Column       |           Type           | Collation | Nullable |                       Default                       
> --------------------+--------------------------+-----------+----------+-----------------------------------------------------
>  id                 | bigint                   |           | not null | nextval('argus_incident_incident_id_seq'::regclass)
> ```

Ok so have investigated this some more and it actually seems to be a small issue with the squashed migration. Firstly if just checkout to master with a blank db and make migrations it makes the same changes for the ID's (except what I added of course), i assume there should be no detected changes. What I think has happened is that originally the ID fields were being autocreated (not defined in model) but then explicitly defined as bigints in 10bc007b672c367d4305340127aae8127cbb5488. So in migration 0001_initial it does ```fields=[("id", models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID"))```  but after being explicitly defined as bigint the alterfield in 0009 does ```field=models.BigAutoField(primary_key=True, serialize=False),```. But in the squashed migration its treating the id's as autocreated even though they are explicitly defined in the model to be bigints. Im not sure how the squashed migration is made but changing
```
('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
to
```
('id', models.BigAutoField(primary_key=True, serialize=False),
```
in the squashed migration stops makemigrations from detecting changes.
DEFAULT_AUTO_FIELD is just AutoField in base.py so doesnt seem to make sense for a BigAutoField to be autocreated.
I suppose this could somehow be an issue on my end or im misunderstanding things haha but this is what I think is going on. I suppose this would need a separate PR to fix the squashed migration?

jladd-geant commented:
Ok have now
- Fixed the issues with the code
- Made a unit test to restart a closed incident then close it again
- Fixed the migration
- Small change to docs

Let me know if you want more tests/changes to what I made, the test does have very similar functionality to the test for reopening i'm not sure if that should be combined.

johannaengland reviewed (changes requested):
Just some nitpicks about naming/function parameters

hmpf commented on tests/incident/test_event.py:
```diff
@@ -72,6 +72,18 @@ def _assert_posting_event_succeeds(self, post_data: dict, client: Client):
         self.assertTrue(self.stateful_incident1.events.filter(pk=response.data["pk"]).exists())
         self.assertEqual(response.data["incident"], self.stateful_incident1.pk)
 
+    def _assert_incident_is_closed_at_timestamp(self, end_event_timestamp):
```
+1

johannaengland committed:
Add new event type `INCIDENT_RESTART`

---------

Co-authored-by: James Ladd <james.ladd@geant.org>

johannaengland committed:
Let API handle new event type `INCIDENT_RESTART`

To make it possible to allow restart events we also allow an incident to have multiple end events
This way a source can restart an incident and end it again

---------

Co-authored-by: James Ladd <james.ladd@geant.org>

johannaengland committed:
Add documentation about `INCIDENT_RESTART` events

---------

Co-authored-by: James Ladd <james.ladd@geant.org>

lunkwill42 commented on src/argus/incident/views.py:
```diff
@@ -506,18 +506,23 @@ def validate_incident_has_no_relation_to_event_type():
 
         if incident.stateful:
-            if event_type in {Event.Type.INCIDENT_START, Event.Type.INCIDENT_END}:
+            if event_type in {Event.Type.INCIDENT_START}:
```
This changes things to allow multiple END events (which fixes other issue we've seen as well): But the change is rather silent (i.e. it is mentioned in the PR on GitHub, but no mention in the commit message - which is what will survive long after GitHub is gone)

johannaengland replied:
Leaving this open to do it when rebasing before merging this

lunkwill42 commented on tests/incident/test_event.py:
```diff
@@ -81,56 +91,94 @@ def test_posting_close_and_reopen_events_properly_changes_stateful_incidents(sel
+            reopen_event_dict, self.stateful_incident1.end_time, self.user1_rest_client
         )
 
+    def test_posting_end_and_restart_events_properly_changes_stateful_incidents(self):
```
The new tests only walk the all-legal transition path (END → RESTART → END, every step valid). The behavior most likely to surprise an integrator is untested: when a **source system** posts an event that doesn't fit the incident's current state, `perform_create` [swallows the validation error](https://github.com/Uninett/Argus/blob/72442dcc563bf1c4cc62a5f8705fb70ee4a6945b/src/argus/incident/views.py#L483-L491) — the event is still *recorded* and the API returns **201**, but `update_incident` is skipped so the incident is left untouched. That record-but-don't-reject bypass is [deliberate](https://github.com/Uninett/Argus/blob/72442dcc563bf1c4cc62a5f8705fb70ee4a6945b/src/argus/incident/views.py#L484-L486), it's new for `INCIDENT_RESTART`, and nothing exercises it. (The same gap exists for `INCIDENT_END`, so it's not a regression here — but RESTART is new and this PR is already open, so this is the moment to lock it down.) Suggest adding:

```python
def test_given_open_incident_when_source_posts_restart_then_records_event_without_state_change(self):
    # An end user posting a state-invalid event gets a 400; a source system does
    # not. Its event is recorded for the audit trail (201) but update_incident is
    # skipped, so the incident is unchanged. Here: RESTART on an already-open
    # incident must be accepted and recorded without re-touching end_time.
    self.assertTrue(self.stateful_incident1.open)
    count_before = self.stateful_incident1.events.count()

    restart_event_dict = self._create_event_dict(Event.Type.INCIDENT_RESTART)
    response = self.source1_rest_client.post(
        self.events_url(self.stateful_incident1), restart_event_dict
    )

    self.assertEqual(response.status_code, status.HTTP_201_CREATED)
    self._assert_incident_is_open(self.stateful_incident1)
    self.stateful_incident1.refresh_from_db()
    self.assertEqual(self.stateful_incident1.events.count(), count_before + 1)
```

lunkwill42 commented on tests/incident/test_event.py:
```diff
@@ -81,56 +91,94 @@ def test_posting_close_and_reopen_events_properly_changes_stateful_incidents(sel
 
         assert_incident_stateless()
-        response = self.user1_rest_client.post(
+        response = self.source1_rest_client.post(
```
Why did this test change which user is posting the event? The test no longer tests what it claims to test because of this: The response from the server is now about source systems not being allowed to post CLOse events; the pre-existing assertion is too weak, it only tests that there is a field error for `type`, not what the error says.  Using a source system as the poster also renders a field error for `type`, but the message is now `A source system cannot post events of type 'CLO'`.

I.e. the test stays incidentally green, though its now testing something entirely different...

johannaengland replied:
I think I got completely confused at this point which events are posted by users and which by source systems...

lunkwill42 commented on tests/incident/test_event.py:
```diff
@@ -81,56 +91,94 @@ def test_posting_close_and_reopen_events_properly_changes_stateful_incidents(sel
         assert_incident_stateless()
 
-        response = self.user1_rest_client.post(
+        response = self.source1_rest_client.post(
```
Same comment about which user account is posting...

lunkwill42 commented on tests/incident/test_event.py:
```diff
@@ -81,56 +91,94 @@ def test_posting_close_and_reopen_events_properly_changes_stateful_incidents(sel
+        response = self.user1_rest_client.post(
+            self.events_url(self.stateless_incident1), self._create_event_dict(Event.Type.INCIDENT_END)
+        )
+        self._assert_response_field_invalid(response, "type")
```
This test is also not testing what it claims to test.  It tries to post end/restart events as a regular user, and hits the `An end user cannot post events of type 'END'` validation error for the `type` field.  So it's not actually testing that the refusal is about attempting to make modifications that are illegal for stateless incidents...

lunkwill42 reviewed (changes requested):
This does solve some pain points, and is nearly there..

There is a manually maintained list of event types also in `user-manual.rst`: This should also be updated before this is merged.

hmpf commented on changelog.d/1502.added.md:
Also mention that there my be more than one END-event.

hmpf commented on src/argus/incident/migrations/0004_alter_event_type.py:
```diff
@@ -0,0 +1,30 @@
+            name="type",
+            field=models.TextField(
+                choices=[
+                    ("STA", "Incident start"),
+                    ("END", "Incident end"),
+                    ("CHI", "Incident change"),
+                    ("RES", "Incident restart"),
+                    ("CLO", "Close"),
+                    ("REO", "Reopen"),
+                    ("ACK", "Acknowledge"),
+                    ("OTH", "Other"),
+                    ("LES", "Stateless"),
```
I've found a way to prevent these noop migrations. Remind me once this one is merged.

johannaengland reviewed (approved):
Reviewing just so that GitHub let's me merge it :grin:
